### ISSUE TITLE
MapFromTree(t *TomlTree) map[string]interface{} ?

### ISSUE THREAD

**Issue by robmuh** [pelletier/go-toml#53]
Would you consider it? Maybe it is there and I'm still missing it. I see `TreeFromMap()` but actually need the reverse, a clean map. From looking at the code there is enough use of custom structs that stripping them all down to bare Go types might be daunting. I like having the syntax specific errors during the parsing, but then need a very generic map. Writing it now and can submit a PR with the additional function perhaps eventually (since I have to have it either way). Feedback appreciated.

**Comment by bep** [comment]
I think what you (and I) want is to export the `values` somehow. There is a `Key()` mehtod, but no `Values()`.

**Comment by pelletier** [comment]
That's definitely something that could be useful in the library. Please open a PR when you have something working, I'd love to review / help with that!

For the record: we have `TomlTree.toToml`, that walks the tree and outputs a TOML representation (so basically performing the opposite operation of `Parse`). Ideally, I'd like to come up with clean way to implement visitors, so that we can build both `toToml` and `toMap` (or whatever we think is clearer) on top of the same generic visitor.

**Comment by bep** [comment]
Ah, OK! Now I see ... What you end up is a tree of `TomlTree`... Not sure I understand the design (parsing TOML to a object representation that still needs some parsing to be fully usable), but it means that this is too much work with hours I haven't got to pull this into Hugo. BurntSushi's library will have to do, I guess.

**Comment by pelletier** [comment]
What do you mean by "fully usable"? The `TomlTree` is expected to be queried using `Get` or `Query` basically.

Do you mind describing your use case, myabe with an example?

I'll start to work on a TomlTree -> Map function that returns only maps and arrays.

**Comment by robmuh** [comment]
Sorry, yes, I meant just simple maps and arrays. a TomlTree -> Map would be ideal. The end result is a data structure that can be used everywhere without the TomlTree dependency, which I think is great for the syntax isolation you have going on. Let me know how interested you are in that, for now, I was going to extract the relevant parsing portions and create a 'Lite' version that does not include those other parts. 

Makes me wonder if there is value in factoring out the state machine parser into its own thing, then the other stuff could be layered onto it different depending on desired results. For example, if someone had another idea about Query and wanted to implement it differently the parsing engine would remain in tact.

**Comment by pelletier** [comment]
That's definitely something I'd like to work on. I have three different approaches to implement that:

##### Approach 1

After the creation of the TomlTree, visit the tree and emit maps and arrays.
That's the simple, easy to write solution. It could even be an interim solution that could provide a good enough solution, and provide a baseline for tests.

##### Approach 2

Define an interface that specifies what to do with each value. That would look something like

```go
type Builder interface {
	String(key string, value string)
	Boolean(key string, value bool)
	Integer(key string, value int64)
	// etc.
}
```

An object  implementing this interface (e.g. `TomlTreeBuilder` or `MapBuilder`) would be passed to the parser, which would in turn call the appropriate function when a new value is encountered during parsing (e.g. `tomlTreeBuilder.String("user.name", "john")`). It is then up to the builder to construct whatever object by handling successive calls to those functions.

##### Approach 3

Pretty much the same as approach 2, but using a channel to communicate between the parser and the builder – in a similar way the lexer and parser communicate. The parser would emit every `(key, value)` pair it encounters during parsing to the channel. The builder would read that channel do whatever it wants with the information, such as building a TomlTree or a map.


I'm still trying to decide between approach 2 and 3. Any suggestion / alternative ideas are welcome!

**Comment by robmuh** [comment]
They all seem great to me. I am partial to the third because it affords the developer the opportunity to make name translations. In my case, for example, I am safe-naming the key names (`some[-_]thing` -> `someThing`, 'some thing' -> 'something') because I want to able to safely flatten the structure into a properties like index suitable for printing to a file and generic viewing. This also plays very nicely with JSONPointer and your Query concept. In other words, being able to transform either the key or value based on logic the developer can add in would be very useful.

**Comment by pelletier** [comment]
I'm thinking about implementing the following pattern:

`Lexer -> Parser -> (Filter 1 -> Filter 2) -> Builder`

arrows being channels, and filters are optional.

There would be (for now) two builders: TomlTreeBuilder and MapBuilder, which construct respectively a TomlTree and a regular map. The idea would be that you can plug in any filters to perform any operations on the (key, value) paris (such as your keys safe-naming), independently of which builder you choose.

How does that sound?

**Comment by robmuh** [comment]
Sounds ideal to me. One thing I might mention is not throwing away comments too early in the process. I would actually like to keep the comments around, in the exact order and position in fact. I think, as a lot of advanced parsers do, they should be considered lexical parts of the file that are optionally discarded by the user of the parser.  This will allow, eventually, write operations that preserve the file exactly, (and which are absent from the other parsers I have seen). If we do it write, we don't have to have a horribly different process for write.

Basically the negative space between things, with or without comments, needs to be treated as a lexical entitiy.

**Comment by pelletier** [comment]
I agree that we should keep the comments and let the builder decide if it wants to do anything with them. I don't know about the value of keeping the blank space as well. Would you have any practical example of what it would be useful for?

@bep just pushed an implementation of `ToMap()` here: https://github.com/pelletier/go-toml/pull/59
You can then do something like

```go
tree, _ := toml.LoadFile('myfile.toml')
myconfig := tree.ToMap() // map[string]interface{}
```

Let me know on the PR if it fits your need and if you find any bug.

**Comment by robmuh** [comment]
That's exactly what I was looking for in the short term.

As for comments, I was hoping to use the extensively to document the TOML file itself and very often people who use comments for such things include one or more blank lines between them for visual separation. One small example of this is some of the Hugo complaints about stripping comments from the front matter. I understand why it happens, just looking for an alternative since I plan to use comments within TOML files as a poor-man's schema documentation for several organized directories of TOML files.

**Comment by robmuh** [comment]
Forked @bep's branch to add some testing directly to it after hitting some panics on the [general example.toml file](https://gist.github.com/robmuh/7966da29024c075349a963840e2298b2) and realizing need more granular tests, which I figured were already there, but turn out to be based on [v0.2.0 set of tests](https://github.com/BurntSushi/toml-test/issues/36) that need to be updated themselves. What do you prefer? To try to keep the 'language agnostic' tests updated there or start our own set here?

**Comment by pelletier** [comment]
@robmuh just tried: it indeed seems like that file is triggering errors. I created an issue https://github.com/pelletier/go-toml/issues/61 to track the efforts to fix those.

**Comment by robmuh** [comment]
Ok, working on an [examples.toml](https://github.com/toml-spec/tests-v0.4.0/blob/master/examples.toml) that is *actually* v0.4.0 compatible. Having a hard time isolating the new cause of the panic. Might be worth a small `tomlv` clone.

**Comment by robmuh** [comment]
Hey @pelletier would you like to track issues for the v0.4.x update somewhere? Been validating @bep's addition and finding small things like the following reference to "key group" which was changed to "table":

```
(*errors.errorString)(0x820335000)((1, 2): unexpected token unclosed key group, was expecting a key group)
```

Been using [go-spew](https://github.com/davecgh/go-spew) to deep dump the structure just for quick validation.

Also, how would you feel about a `messages.go` refactoring PR if all moved all the text messages into constants in that file, (which could be internationalized better later as well)? Is that worth considering?

**Comment by robmuh** [comment]
This seems to fail:
Another heads up:

```toml
date4 = 1979-05-27T07:32:00
```

Fails but supported on the main page, although not in the v0.4.0 release.

**Comment by robmuh** [comment]
Ok, this one looks legit, it's cut and paste from the v0.4.0 page:

```toml
key1 = """
Roses are red
Violets are blue"""
```
```
(2, 1): unescaped control character U+000A
```

**Comment by robmuh** [comment]
Looks like inline-tables have a problem as well:

```toml
name = { first = "Tom", last = "Preston-Werner" }
points = { x = 1, y = 2 }
```

At first I thought these were not v0.4.0 but the are.

**Comment by robmuh** [comment]
That wraps up testing everything from the v0.4.0 page and a little from whatever comes next. Turns out this package makes a good [`tomll`](http://github.com/toml-spec/tomll) list tool, which I used against [these tests](http://github.com/toml-spec/tests-v0.4.0). I will be using @bep's `ToMap()` to finish up the stuff I needed and just make sure not to use double-quoted long strings and inline-maps for now. Thanks. Let me know if there is more I could help with.

**Comment by pelletier** [comment]
@robmuh 

https://github.com/pelletier/go-toml/issues/53#issuecomment-213073178
https://github.com/pelletier/go-toml/issues/53#issuecomment-213230954
Those two have been fixed by https://github.com/pelletier/go-toml/pull/62.

> Hey @pelletier would you like to track issues for the v0.4.x update somewhere?

Yes! There is a Github milestone for that: https://github.com/pelletier/go-toml/milestones/Future%20TOML

> Also, how would you feel about a messages.go refactoring PR if all moved all the text messages into constants in that file, (which could be internationalized better later as well)? Is that worth considering?

I haven't really considered internationalization yet. I have no idea what are the best practices for that in Go world. Maybe a GNU Gettext implementation could be used? As for a file with all the messages: I'm not in favor. Having the error message where it happens allow you to jump directly to the code, instead of having to go through the messages.go file.

> [Dates] seem to fail

Created https://github.com/pelletier/go-toml/issues/63 to track support of those formats.

> Looks like inline-tables have a problem as well.

Those *are* in the 0.4.0 release, and are supposed to be supported, so that's a bug. Filed https://github.com/pelletier/go-toml/issues/64.

> I will be using @bep's ToMap() to finish up the stuff I needed

Do you mean this function? https://github.com/pelletier/go-toml/pull/59 I'm about to merge it in master. If you're not referring to that one, would you mind elaborating on the issues you're encountering with it?

**Comment by pelletier** [comment]
Closing this as https://github.com/pelletier/go-toml/pull/59 has been merged. Let's continue the discussions on specialized Github issues.

**Comment by robmuh** [comment]
All sounds good. I was actually saying that @bep's function in #59 works for me, (except for the panic and inline tables), so I agree with the merge and it will make using `go-toml` easier now because I won't have to fork and switch the branch as I did for testing, (which made me really miss a `go get` convention for GitHub that let's you import a specific branch).

I will retest the panic on the merged new version. Maybe the branch or my setup had something to do with it.

**Comment by pelletier** [comment]
By the way you can use gopkg.in if you want to pull from a specific branch. Definitely not idea, but may save the day until with have something baked in go.
